doc/source/DBEP/DBEP03.rst

```diff
@@ -0,0 +1,59 @@
+==============================================================
+  Support bulk event / datum tables as first-class documents
+==============================================================
+
+.. contents::
+   :local:
+
+
+Status
+======
+
+
+**Discussion**
+
+Branches and Pull requests
+==========================
+
+
+Abstract
+========
+
+On both the ingest and access sides it may not be practical, for
+performance reasons, to access *Event* and *Datum* on a
+document-by-document basis.  Within a given stream and
+*Resource* *Event* and *Datum* documents are consistent and thus can
+be represented as dense tables.
+
+We currently have a *bulk_event* schema, however it is an array of
+*Event* not a table and does not vectorize in the correct ways.
+
+
+Detailed description
+====================
+
+For both *EventTable* and *DatumTable* the schema is identical to the single
+version except in place of scaler values there is an array of scaler values.
+These can be represented as:
+
+- nested `dict` with `numpy.array` or `list` at the bottom
+- a `dict` of `pandas.DataFrame` with simple column index and a shared row index
+- a `pandas.DataFrame` with a MultiIndex
+
```

Comment:
On the access side, getting back a single array (or array-like object) containing all event data for a particular field would be an amazing feature. It should be noted though that for long scans I wouldn't want to transmit all frames at once when accessing that table. While I would want an object describing the shape, min, max, and size of the array to represent the contents, I might only access a subset of that array (typically slicing on the event/frame axis). Can I also have the option to get back an array-like object with 'lazy' access to each frame?

Reply:
Yes to all of those :smile: 

 - we should be able ask for the tables as chunks (ex 100 rows at a time).  This fits very well with what apache arrow supports.
 - we should be able to ask for any sub-set of the columns.  we already have that on some of the DB api, but needs to be everywhere.  There is also a lot of space for optimization in this area.
 - the shape should be known from the descriptor (the type is _not_ well known past 'number', that is something else that we need to look into improving), but I don't think we compute any statistics on the raw-data (but that maybe something we should be grabbing at collection time).
 - the `data` method on `Header` will give you a generator of a specific field (https://nsls-ii.github.io/databroker/_as_gen/databroker.Header.data.html#databroker.Header.data)
 - There is already a way to get a `pims` object which will do lazy asset registry (nee filestore) retrievals and several proto-type dask-based

Reply:
The dask prototype is [here](https://github.com/NSLS-II-CHX/eiger-io/blob/master/eiger_io/fs_handler_dask.py) but it's a convoluted example for a couple reasons (the Eiger data format is unusual, and we're working around some historical baggage here...). In principle, dask should win us the 'lazy along any axis' feature, modulo what HDF5 can support.